service/udp.go

```diff
@@ -71,136 +78,131 @@ func findAccessKeyUDP(clientIP netip.Addr, dst, src []byte, cipherList CipherLis
+
+	defer func() {
+		debugUDP(l, "Done")
+		assocMetrics.AddClose()
+	}()
 
+	var targetConn net.PacketConn
+	var cryptoKey *shadowsocks.EncryptionKey
+
+	readBufLazySlice := readBufPool.LazySlice()
+	readBuf := readBufLazySlice.Acquire()
+	defer readBufLazySlice.Release()
 	for {
-		clientProxyBytes, clientAddr, err := clientConn.ReadFrom(cipherBuf)
-		if errors.Is(err, net.ErrClosed) {
+		select {
+		case <-ctx.Done():
+			break
+		default:
+		}
+		clientProxyBytes, err := clientConn.Read(readBuf)
+		if errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF) {
+			debugUDP(l, "Client closed connection")
 			break
 		}
+		pkt := readBuf[:clientProxyBytes]
+		debugUDP(l, "Outbound packet.", slog.Int("bytes", clientProxyBytes))
 
 		var proxyTargetBytes int
-		var targetConn *natconn
-
-		connError := func() (connError *onet.ConnectionError) {
-			defer func() {
-				if r := recover(); r != nil {
-					slog.Error("Panic in UDP loop: %v. Continuing to listen.", r)
-					debug.PrintStack()
-				}
-			}()
-
-			// Error from ReadFrom
-			if err != nil {
-				return onet.NewConnectionError("ERR_READ", "Failed to read from client", err)
-			}
-			defer slog.LogAttrs(nil, slog.LevelDebug, "UDP: Done", slog.String("address", clientAddr.String()))
-			debugUDPAddr(h.logger, "Outbound packet.", clientAddr, slog.Int("bytes", clientProxyBytes))
 
-			cipherData := cipherBuf[:clientProxyBytes]
+		connError := func() *onet.ConnectionError {
 			var payload []byte
 			var tgtUDPAddr *net.UDPAddr
-			targetConn = nm.Get(clientAddr.String())
 			if targetConn == nil {
-				ip := clientAddr.(*net.UDPAddr).AddrPort().Addr()
+				ip := clientConn.RemoteAddr().(*net.UDPAddr).AddrPort().Addr()
 				var textData []byte
-				var cryptoKey *shadowsocks.EncryptionKey
+				var keyID string
+				textLazySlice := readBufPool.LazySlice()
 				unpackStart := time.Now()
-				textData, keyID, cryptoKey, err := findAccessKeyUDP(ip, textBuf, cipherData, h.ciphers, h.logger)
+				textData, keyID, cryptoKey, err = findAccessKeyUDP(ip, textLazySlice.Acquire(), pkt, h.ciphers, h.logger)
 				timeToCipher := time.Since(unpackStart)
+				textLazySlice.Release()
 				h.ssm.AddCipherSearch(err == nil, timeToCipher)
 
 				if err != nil {
 					return onet.NewConnectionError("ERR_CIPHER", "Failed to unpack initial packet", err)
 				}
+				assocMetrics.AddAuthentication(keyID)
 
 				var onetErr *onet.ConnectionError
 				if payload, tgtUDPAddr, onetErr = h.validatePacket(textData); onetErr != nil {
 					return onetErr
 				}
 
-				udpConn, err := h.targetListener.ListenPacket(context.Background())
+				// Create the target connection.
+				targetConn, err = h.targetListener.ListenPacket(ctx)
 				if err != nil {
 					return onet.NewConnectionError("ERR_CREATE_SOCKET", "Failed to create a `PacketConn`", err)
 				}
-
-				targetConn = nm.Add(clientAddr, clientConn, cryptoKey, udpConn, keyID)
+				l = l.With(slog.Any("ltarget", targetConn.LocalAddr()))
```

Comment:
Maybe something more explicit?
```suggestion
				l = l.With(slog.Any("tgtListener", targetConn.LocalAddr()))
```

It was not clear what `llistener` was